detectors/node/opentelemetry-resource-detector-container/src/detectors/ContainerDetector.ts

```diff
@@ -94,22 +81,70 @@ export class ContainerDetector implements Detector {
       .trim()
       .split('\n')
       .find(s => s.includes(this.HOSTNAME));
-    const containerIdStr = str
-      ?.split('/')
-      .find(s => s.length === this.CONTAINER_ID_LENGTH);
-    return containerIdStr || '';
+
+    const strArray = str?.split('/') ?? [];
+    for (let i = 0; i < strArray?.length; i++) {
+      if (
+        strArray[i] === this.MARKING_PREFIX &&
+        strArray[i + 1]?.length === this.CONTAINER_ID_LENGTH
+      ) {
+        return strArray[i + 1];
+      }
+    }
+    return '';
+  }
+
+  private truncatePrefix(lastSection: string, prefix: string): string {
+    return lastSection.substring(prefix.length);
   }
 
+  protected extractContainerIdFromLine(line: string): string | null {
+    if (line) {
+      const sections = line.split('/');
+      if (sections.length > 1) {
+        let lastSection = sections[sections.length - 1];
+
+        // Handle containerd v1.5.0+ format with systemd cgroup driver
+        const colonIndex = lastSection.lastIndexOf(':');
+        if (colonIndex !== -1) {
+          lastSection = lastSection.substring(colonIndex + 1);
+        }
+
+        // Truncate known prefixes from the last section
+        if (lastSection.startsWith(this.CRIO)) {
+          lastSection = this.truncatePrefix(lastSection, this.CRIO);
+        } else if (lastSection.startsWith(this.DOCKER)) {
+          lastSection = this.truncatePrefix(lastSection, this.DOCKER);
+        } else if (lastSection.startsWith(this.CRI_CONTAINERD)) {
+          lastSection = this.truncatePrefix(lastSection, this.CRI_CONTAINERD);
+        }
+        // Remove anything after the first period
+        if (lastSection.includes('.')) {
+          lastSection = lastSection.split('.')[0];
```

Comment:
the original code is using the last dot:

```js
        let startIdx = lastSection.lastIndexOf('-');
        let endIdx = lastSection.lastIndexOf('.');

        startIdx = startIdx === -1 ? 0 : startIdx + 1;
        if (endIdx === -1) {
          endIdx = lastSection.length;
        }
        if (startIdx > endIdx) {
          continue;
        }
```

Can you elaborate about the motivation for this change, perhaps also include some examples of what this`section` might be formatted like to give more context for the review and future maintainers?

Reply:
So the motivation for this change is as follows:

1. Add more cases to be handled ( more prefixes ) - previous case only handled a specific case
2. Adding more filter - knowing that v1 id is hexadecimal in nature - adding that check 
3. The new logic is essentially doing very similar but with more checks
4. Added a lot of tests and example in utils.test.ts